### ISSUE TITLE
H2O benchmarks are incorrect & misleading

### ISSUE THREAD

**Issue by ledell** [tomaztk/SLODUG2017#1]
There are many issues with this benchmark against H2O (`R_digit_recognizer_3_V2.R`).  Here are a few things I see just off the top of my head:
- You don't use a test set in the H2O demo.  How can you compare test set performance between two algos when without a test set?
- The networks are not the same, so this is not an apples-to-apples comparison of the algos.
- A benchmark on a single set of parameters with the goal of producing a generalizable result is not a benchmark.  You will need to tune each algo/implementation separately.  If you want this to be a fair comparison, based on speed-to-train, then you'd need to give each algo the same amount of time to perform the tuning / hyperparameter search and then compare the test set performance of the best model from each implementation.  (The same recommendations apply to the xgboost benchmark as well).  

Even if you use the default parameters for H2O (without any tuning), you can get better performance than what you are reporting.  R code example:

```r
library(h2o)
h2o.init(nthreads = -1)

# Load MNIST train/test data
train_file <- "https://h2o-public-test-data.s3.amazonaws.com/bigdata/laptop/mnist/train.csv.gz"
test_file <- "https://h2o-public-test-data.s3.amazonaws.com/bigdata/laptop/mnist/test.csv.gz"
train <- h2o.importFile(train_file)
test <- h2o.importFile(test_file)
y <- "C785"
x <- setdiff(names(train), y)

# Encode the response column as categorical for multinomial classification
train[,y] <- as.factor(train[,y])
test[,y] <- as.factor(test[,y])

# Train Deep Learning model with all default parameters
model <- h2o.deeplearning(x = x, 
                          y = y, 
                          training_frame = train,
                          distribution = "multinomial")

# Evaluate performance on a test set
perf <- h2o.performance(model, newdata = test)
h2o.confusionMatrix(perf)
```
Gives about 96.8% overall accuracy without even trying.

```r
> h2o.confusionMatrix(perf)
Confusion Matrix: vertical: actual; across: predicted
         0    1    2   3   4   5   6    7    8   9  Error           Rate
0      962    0    1   1   2   6   1    1    6   0 0.0184 =     18 / 980
1        0 1127    3   0   0   1   1    1    2   0 0.0070 =    8 / 1,135
2        5    2  989   6   2   2   5    7   13   1 0.0417 =   43 / 1,032
3        1    0    8 972   0   5   0   11   10   3 0.0376 =   38 / 1,010
4        2    0    4   1 958   0   4    5    2   6 0.0244 =     24 / 982
5        4    0    1   4   1 866   2    3   10   1 0.0291 =     26 / 892
6        6    5    3   0  13   9 916    0    6   0 0.0438 =     42 / 958
7        1    5    8   2   1   0   0  996    6   9 0.0311 =   32 / 1,028
8        7    0    4   0   3  10   1    7  942   0 0.0329 =     32 / 974
9        7    4    0   6  12   4   0   12   12 952 0.0565 =   57 / 1,009
Totals 995 1143 1021 992 992 903 930 1043 1009 972 0.0320 = 320 / 10,000
> 1-0.0320
[1] 0.968
```

Also note: The parameters you have on your blog post do not match the code in this file: `R_digit_recognizer_3_V2.R` 

In the interest of honest benchmarking, it would be great if you could fix the issues with the benchmark or update your [blog post](https://tomaztsql.wordpress.com/2017/02/20/rxneuralnet-vs-xgboost-vs-h2o/) to indicate that a simple comparison like this should not be considered to be a valid benchmark of one piece of software over another.  Most experienced machine learning folks will be able to understand that this is not a useful comparison, but those who are new to machine learning will not know the difference and could be mislead.

**Comment by tomaztk** [comment]
Hi,

Thank you for the comment and the insights. Much appreciated.

First of all, I am a big fan of h2o R package, and also the Python package. And I love to use it. So, the idea was not to mislead anyone - hence the code is always included - or discredit anyone. It was just a curiosity performance check.

In the test, I was focusing on performance/time test; since h2o and revoscaler are very good in accuracy and predictions. And as graphs suggests, there is time comparisons.

I was using same MNIST dataset, and the networks, should be  the same. But again, thanks for pointing that out, I will check the code and results again.  Based on my blog post I got 95,8% accur,  you got 96,8 accur., which is even better and I will rerun the test.

Thank, Tomaž

**Comment by ledell** [comment]
It looks like you are using a [Convolutional Neural Network (CNN)](https://en.wikipedia.org/wiki/Convolutional_neural_network) in the `rxNeuralNet()` function.  H2O's `h2o.deeplearning()` function implements the [Multilayer Perceptron (MLP)](https://en.wikipedia.org/wiki/Multilayer_perceptron), so unless I am missing something, these are not the same networks at all.  

Why did you choose this particular set of values for the network hyperparameters?  That doesn't make a lot of sense to me.  If you check out the [H2O Deep Learning Vignette](http://docs.h2o.ai/h2o/latest-stable/h2o-docs/booklets/DeepLearningBooklet.pdf), we use MNIST as an example, which will show you how to get good performance from H2O rather than relying on a (seemingly) random choice of hyperparameters.  Here are a few other resources you can use: [H2O Deep Learning Performance Tips blog post](http://blog.h2o.ai/2015/08/deep-learning-performance-august/) and H2O Deep Learning [documentation](http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/deep-learning.html) and [tutorial](http://docs.h2o.ai/h2o-tutorials/latest-stable/tutorials/deeplearning/index.html).

On the bright side, I'm glad to hear that you enjoy using H2O.  :-)

**Comment by tomaztk** [comment]
Hi,

I wasn't aiming at the CNN but more into Perceptron area. I have to check which version of rx I was using. With versioin 9.0.3 they have introduced CNN as well, but prior to this, parceptrons must be available. I wanted to have more similar NN to what is  available in the Azure ML cloud, rather than doing CNN and to compare the training and scoring time on-prem vs. in the cloud. EC2 and Azure ML both charge based on training time. But this is also something on the list to check.

Thanks for the resources blog post. I went through the vignettes but additional documentation will be a absolute helper.  

One thing I was asking on h2o team on twitter was, the question if you can handle the parallel and distributed computation (on-prem), especially when data can not fit into the memory. 
Best, Tomaž

**Comment by ledell** [comment]
> One thing I was asking on h2o team on twitter was, the question if you can handle the parallel and distributed computation (on-prem), especially when data can not fit into the memory.

Do you mean, can H2O be used on-prem instead of in the cloud?  There has never been a cloud-based H2O-as-a-service, so everyone using H2O is using it either on-prem or in a cloud service where they manage the nodes themselves (e.g. Amazon EC2, etc).  All computations are distributed in H2O regardless of whether you are using your laptop (in this case, distributed just across the cores on your laptop) or a mulitnode cluster.   Hopefully that answers your question.
